rollup/main/default/classes/RollupRelationshipFieldFinder.cls

```diff
@@ -0,0 +1,275 @@
+    private Boolean isAbortedEarly = false;
+
+    private final Map<Id, SObject> lookupIdToFinalRecords = new Map<Id, SObject>();
+    private Map<Id, List<Id>> lookupIdMap = new Map<Id, List<Id>>();
+    private final Map<Id, List<Id>> hierarchy = new Map<Id, List<Id>>();
+    private final RollupRelationshipFieldFinder finder;
+
+    public Traversal(RollupRelationshipFieldFinder finder) {
+      this.finder = finder;
+    }
+
+    public Boolean getIsFinished() {
+      return this.isFinished;
+    }
+
+    public SObject retrieveParent(Id descendantId) {
+      return this.lookupIdToFinalRecords.get(descendantId);
+    }
+
+    public List<SObject> getAllParents() {
+      if (this.isAbortedEarly) {
+        return new List<SObject>();
+      }
+      // not ideal, but because multiple parents can be tied to different descendants ...
+      return new List<SObject>(new Set<SObject>(this.lookupIdToFinalRecords.values()));
+    }
+
+    public void recommence() {
+      this.finder.getParents(this.finder.recommencementRecords);
+    }
+
+    public Map<String, List<SObject>> getParentLookupToRecords() {
+      Map<String, List<SObject>> parentToLookupRecords = new Map<String, List<SObject>>();
+      if (this.isAbortedEarly) {
+        return parentToLookupRecords;
+      }
+      for (SObject record : this.finder.records) {
+        SObject parentRecord = this.retrieveParent(record.Id);
+        if (parentToLookupRecords.containsKey(parentRecord.Id)) {
+          parentToLookupRecords.get(parentRecord.Id).add(record);
+        } else {
+          parentToLookupRecords.put(parentRecord.Id, new List<SObject>{ record });
+        }
+      }
+      return parentToLookupRecords;
+    }
+
+    public Boolean isUltimatelyReparented(SObject record, String relationshipFieldName) {
+      Id currentLookupId = (Id) record.get(relationshipFieldName);
+      Id oldLookupId = (Id) (this.finder.oldRecords.containsKey(record.Id)
+        ? this.finder.oldRecords.get(record.Id).get(relationshipFieldName)
+        : currentLookupId);
+      if (currentLookupId == oldLookupId) {
+        return false;
+      } else if (currentLookupId == null || oldLookupId == null) {
+        // this is pretty cut and dry. if we are moving from having a lookup to not having one, or vice versa, it's a reparenting
+        return true;
+      }
+      CombinedHierarchy combinedHierarchy = this.getHierarchy(record, relationshipFieldName);
+      if (combinedHierarchy.currentHierarchy?.size() > 0 && combinedHierarchy.oldHierarchy?.size() > 0) {
+        // the last Ids present in the chain have to match, otherwise it's a reparenting
+        return combinedHierarchy.currentHierarchy[combinedHierarchy.currentHierarchy.size() - 1] !=
+          combinedHierarchy.oldHierarchy[combinedHierarchy.oldHierarchy.size() - 1];
+      } else {
+        // if there was only one hop, we can just compare the Ids
+        return currentLookupId != oldLookupId;
```

Comment:
I should update this note - it has to go last. The hierarchical comparison needs to be allowed to run first to avoid reporting false positives. Take a look at the comment block I made around line 235 for the full explanation!